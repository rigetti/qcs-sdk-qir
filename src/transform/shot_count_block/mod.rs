mod pattern;
pub(crate) mod qir;
pub(crate) mod quil;

pub(crate) const PARAMETER_MEMORY_REGION_NAME: &str = "__qir_param";
