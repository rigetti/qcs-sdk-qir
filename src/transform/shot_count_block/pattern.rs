// Copyright 2022 Rigetti Computing
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// This module concerns itself with the handling of patterns in spans of instructions.
use either::Either;
use inkwell::{
    basic_block::BasicBlock,
    values::{
        BasicValue, BasicValueEnum, FloatValue, FunctionValue, InstructionOpcode, InstructionValue,
    },
};
use lazy_static::lazy_static;
use log::{debug, info};
use quil_rs::{
    expression::Expression,
    instruction::{GateModifier, MemoryReference, Qubit},
};
use regex::Regex;
use std::{
    collections::{hash_map::DefaultHasher, HashMap},
    convert::TryFrom,
    hash::{Hash, Hasher},
};

use crate::{
    context::QCSCompilerContext,
    interop::instruction::{
        get_called_function_name, get_qis_function_arguments, integer_value_to_u64,
        operand_to_integer, OperationArgument,
    },
};

use super::PARAMETER_MEMORY_REGION_NAME;

/// Each pattern matching function, if it finds a match, returns a result tuple including:
/// * data about the pattern which was matched.
/// * what the next instruction following the pattern is, if any.
type PatternResult<'ctx, T> = Option<(Option<InstructionValue<'ctx>>, T)>;

/// A `ShotCountPatternMatchContext` accumulates state as it scans a QIR program.
/// It is used to infer a shot count from a fixed count loop wrapping some number of
/// quantum and classical instructions within a basic block.
///
/// The pattern it looks for is the following:
///
/// * loop start and shot count - see [`shot_count_loop_start`].
/// * any number of the following:
///   * quantum instructions, which it earmarks for removal from the program, transpiles to Quil,
///     and appends to its running Quil program - see [`quantum_instruction`]
///   * classical instructions, which are ignored and left in place.
/// * a shot count increment and branch instruction - see [`shot_count_loop_end`].
#[derive(Debug, Default)]
pub(crate) struct ShotCountPatternMatchContext<'ctx> {
    // The instruction used to initialize the shot count value
    // pub loop_initializer: Option<InstructionValue<'ctx>>,
    /// Hash of the instruction used to initialize the shot count value.
    /// By using the hash, we don't need to maintain a reference to the instruction itself.
    pub initial_instruction: Option<InstructionValue<'ctx>>,

    /// Hash of the instruction used to initialize the shot count value.
    /// By using the hash, we don't need to maintain a reference to the instruction itself.
    pub initial_instruction_hash: Option<u64>,

    /// The quil program transpiled from quantum intrinsics
    pub quil_program: quil_rs::Program,

    /// The shot count inferred from the loop instructions
    pub shot_count: Option<u64>,

    /// A list of instructions to remove from the program (for substitution with Quil)
    pub instructions_to_remove: Vec<InstructionValue<'ctx>>,

    /// The terminating instruction to append to the BasicBlock in place of a conditional
    /// branch on shot count
    pub next_basic_block: Option<BasicBlock<'ctx>>,

    /// Mapping of (read_result *Result index)->(ro memory region index)
    pub read_result_mapping: HashMap<u64, u64>,

    /// Pairings of (readout buffer index/offset) with the instruction which stores that readout value.
    pub readout_instruction_mapping: Vec<(u64, InstructionValue<'ctx>)>,

    // All FloatValues used as instruction parameters. Indices within this Vec map to indices within the Quil
    // MemoryRegion used to read the values at runtime.
    pub parameters: Vec<FloatValue<'ctx>>,

    /// Whether or not to prepend a RESET instruction to the program to actively reset all qubits on each shot
    pub use_active_reset: bool,
}

impl<'ctx> ShotCountPatternMatchContext<'ctx> {
    /// If the program contains any executable instructions (gates, pulses, etc) and a shot count has been inferred,
    /// return that information; otherwise, return `None` indicating that the pattern was not matched.
    pub fn get_program_data(&self) -> Option<(&quil_rs::Program, u64)> {
        if let Some(shots) = self.shot_count {
            if self.quil_program.to_instructions(false).is_empty() {
                None
            } else {
                Some((&self.quil_program, shots))
            }
        } else {
            None
        }
    }

    /// Build the pattern context from a basic block.
    ///
    /// # Arguments
    ///
    /// * `context`: overall compiler context
    /// * `basic_block`: the subject block to be searched for the pattern
    /// * `visited_functions`: list of function names which have already been transpiled. Used to prevent recursion loops.
    /// * `function_call_callback`: callback to be invoked when a function call is found within the block, in order to recursively
    ///    transpile an LLVM module.
    pub fn from_basic_block(
        context: &mut QCSCompilerContext<'ctx>,
        basic_block: BasicBlock<'ctx>,
        visited_functions: &[&str],
        function_call_callback: fn(
            &mut QCSCompilerContext<'ctx>,
            FunctionValue<'ctx>,
            &[&str],
        ) -> eyre::Result<()>,
    ) -> eyre::Result<Self> {
        let mut next_instruction = basic_block.get_first_instruction();
        let mut pattern_context = ShotCountPatternMatchContext::default();

        info!(
            "starting transpile: block {}",
            basic_block.get_name().to_str().unwrap()
        );

        while let Some(instruction) = next_instruction {
            if instruction.get_opcode() == InstructionOpcode::Call {
                // TODO: handle callbr?
                if let Some(Either::Left(BasicValueEnum::PointerValue(pointer_value))) =
                    instruction.get_operand(0)
                {
                    let function_name = pointer_value.get_name().to_str().unwrap();
                    if !visited_functions.contains(&function_name) {
                        if let Some(function) = context.module.get_function(function_name) {
                            let mut visited_functions = Vec::from(visited_functions);
                            visited_functions.push(function_name);
                            function_call_callback(context, function, &visited_functions)?;
                            next_instruction = instruction.get_next_instruction();
                            continue;
                        }
                    }
                }
            }

            // If we haven't yet found the loop start...
            if pattern_context.initial_instruction.is_none() {
                // Check if we've found it in this instruction. If not, continue on to the next instruction until we do find it.
                // FIXME: ensure we encounter this first (i.e. the pattern must be matched in order)
                if let Some((pattern_instruction, _)) =
                    shot_count_loop_start(&mut pattern_context, instruction)
                {
                    debug!("matched shot count start: {:?}", instruction);
                    next_instruction = pattern_instruction;
                    continue;
                }
            } else if let Some((pattern_instruction, _)) =
                quantum_instruction(context, &mut pattern_context, instruction)
            {
                debug!("matched quantum instruction: {:?}", instruction);
                next_instruction = pattern_instruction;
                continue;
            } else if let Some((_, _)) =
                shot_count_loop_end(context, &mut pattern_context, instruction)
            {
                debug!("matched shot count end: {:?}", instruction);
                break;
            }

            next_instruction = instruction.get_next_instruction();
        }

        Ok(pattern_context)
    }
}

/// Match the initial instruction of a shot-count loop. This may take one of the following forms:
///
/// * a `phi` instruction with two branches, where one of the branches is the name of the current basic block
///   and the other branch sets a constant value of `1`
///
/// Example:
///
/// ```llvm
/// %116 = phi i64 [ %119, %body__1.i15.i23 ], [ 1, %body__1.i12.i19 ]
/// ```
///
/// If matched, this function returns the variable name assigned by the `phi` operand, for use in identifying the end of the loop,
/// as well as the number of shots.
pub(crate) fn shot_count_loop_start<'a, 'ctx>(
    pattern_context: &'a mut ShotCountPatternMatchContext<'ctx>,
    instruction: InstructionValue<'ctx>,
) -> PatternResult<'ctx, ()> {
    match instruction.get_opcode() {
        inkwell::values::InstructionOpcode::Phi => {
            // println!("Phi instruction: {:?}", instruction);
            // TODO: figure out references here to prevent this error:
            // While deleting: i64 %
            // Use still stuck around after Def is destroyed:  <badref> = phi i64 [ <badref>, %body ], [ 1, blockaddress
            // pattern_context.variable_name = Some(instruction.clone());
            pattern_context.initial_instruction = Some(instruction);
            // pattern_context.instructions_to_remove.push(instruction);

            let mut hasher = DefaultHasher::new();
            instruction.hash(&mut hasher);
            pattern_context.initial_instruction_hash = Some(hasher.finish());
            Some((instruction.get_next_instruction(), ()))
        }
        _ => None,
    }
}

/// Match the final instructions of a shot-count loop, which increment the shot count,
/// test for equality to the number of shots, and then rev
///
/// Example:
///
/// ```llvm
/// %119 = add nuw nsw i64 %116, 1
/// %120 = icmp ult i64 %116, 1000
/// br i1 %120, label %body__1.i15.i23, label %body__1.i18.i27
/// ```
///
/// These three instructions must immediately follow one another as depicted here.
///
/// If matched, this function returns the shot count.
pub(crate) fn shot_count_loop_end<'a, 'ctx>(
    context: &QCSCompilerContext,
    pattern_context: &'a mut ShotCountPatternMatchContext<'ctx>,
    instruction: InstructionValue<'ctx>,
) -> PatternResult<'ctx, ()> {
    match instruction.get_opcode() {
        inkwell::values::InstructionOpcode::Add => {
            // We only want to match spans starting with an add of constant 1 to the same register
            // initialized at the beginning of the block, i.e. `add nuw nsw i64 %0, 1`
            let shot_count_increment_is_1 = instruction
                .get_operand(1)
                .and_then(operand_to_integer)
                .and_then(|integer| integer_value_to_u64(context, integer))
                .map_or(false, |int_value| int_value == 1);

            if shot_count_increment_is_1 {
                // Here: test that the target of the instruction is same as the shot count start variable
                if let Some(next_instruction) = instruction.get_next_instruction() {
                    if next_instruction.get_opcode() == inkwell::values::InstructionOpcode::ICmp {
                        // Here: test that the first operand is a constant; extract the shot count
                        // and save that into the pattern context
                        if let Some(Either::Left(BasicValueEnum::IntValue(_))) =
                            next_instruction.get_operand(0)
                        {
                            // Test that the first operand is the shot count variable (the initial Phi instruction)
                            let matches_shot_count_loop_start =
                                pattern_context.initial_instruction.map_or(false, |instr| {
                                    next_instruction
                                        .get_operand_use(0)
                                        .unwrap()
                                        .get_used_value()
                                        .left()
                                        .unwrap()
                                        .as_instruction_value()
                                        == Some(instr)
                                });

                            if matches_shot_count_loop_start {
                                let operand = next_instruction.get_operand(1);
                                if let Some(Either::Left(BasicValueEnum::IntValue(operand_value))) =
                                    operand
                                {
                                    let shot_count = u64::try_from(
                                        operand_value.get_sign_extended_constant().unwrap(),
                                    )
                                    .expect("shot count value must be non-negative");

                                    if let Some(final_instruction) =
                                        next_instruction.get_next_instruction()
                                    {
                                        // Test that it's branching on the correct instruction value (the result of the comparison)
                                        let branching_on_loop_variable = next_instruction
                                            .get_first_use()
                                            == final_instruction.get_operand_use(0);

                                        if final_instruction.get_opcode() == InstructionOpcode::Br
                                            && branching_on_loop_variable
                                        {
                                            pattern_context.shot_count = Some(shot_count);

                                            if let Some(Either::Right(next_basic_block)) =
                                                final_instruction.get_operand(1)
                                            {
                                                pattern_context.next_basic_block =
                                                    Some(next_basic_block);
                                            }
                                        } else {
                                            panic!("expected branch instruction to end shot count block, got {:?}", final_instruction);
                                        }
                                    } else {
                                        panic!("expected a branch instruction to end shot count block, none present");
                                    }
                                } else {
                                    panic!("expected integer operand, got {:?}", operand);
                                }
                            } else {
                                panic!(
                                    "expected\n{:?}\nto equal\n{:?}",
                                    next_instruction
                                        .get_operand_use(0)
                                        .unwrap()
                                        .get_used_value()
                                        .left()
                                        .unwrap()
                                        .as_instruction_value(),
                                    pattern_context.initial_instruction
                                );
                            }
                        }
                    }

                    return Some((next_instruction.get_next_instruction(), ()));
                }
            }
            None
        }
        _ => None,
    }
}

macro_rules! match_qis_argument {
    ($variant:ident, $arguments:expr, $index:expr, $function_name:expr) => {{
        use OperationArgument::*;

        match $arguments.get($index) {
            Some($variant(contents)) => contents,
            other => {
                panic!(
                    "expected argument {} of {} to be of type {:?}; got {:?}",
                    $index,
                    $function_name,
                    stringify!($variant),
                    other
                )
            }
        }
    }};
}

/// Given a `FloatValue` which may be the parameter of a QIS intrinsic call, return the Quil Expression
/// which will be used to store its value at execution time.
fn get_quil_parameter_expression<'ctx>(
    pattern_context: &mut ShotCountPatternMatchContext<'ctx>,
    float_value: FloatValue<'ctx>,
) -> Expression {
    let index = get_quil_parameter_index(pattern_context, float_value);
    Expression::Address(MemoryReference {
        name: String::from(PARAMETER_MEMORY_REGION_NAME),
        index: index as u64,
    })
}

/// Given a `FloatValue` to be used as the parameter to a gate, return the index within the
/// Quil `MemoryReference` that should be used to store this parameter's value.
pub(crate) fn get_quil_parameter_index<'ctx>(
    pattern_context: &mut ShotCountPatternMatchContext<'ctx>,
    float_value: FloatValue<'ctx>,
) -> usize {
    if let Some(index) = pattern_context
        .parameters
        .iter()
        .position(|el| el == &float_value)
    {
        index
    } else {
        pattern_context.parameters.push(float_value);
        pattern_context.parameters.len() - 1
    }
}

#[allow(clippy::too_many_arguments)]
fn add_gate_instruction<'ctx>(
    pattern_context: &mut ShotCountPatternMatchContext<'ctx>,
    arguments: &[OperationArgument<'ctx>],
    function_name: &str,
    name: &str,
    adjoint: bool,
    controlled: bool,
    parameter_count: usize,
    qubit_count: usize,
) {
    let parameters = (0..parameter_count)
        .map(|arg_index| {
            let float_value = *match_qis_argument!(Parameter, arguments, arg_index, function_name);
            get_quil_parameter_expression(pattern_context, float_value)
        })
        .collect();
    let qubits = (parameter_count..parameter_count + qubit_count)
        .map(|arg_index| {
            quil_rs::instruction::Qubit::Fixed(*match_qis_argument!(
                Qubit,
                arguments,
                arg_index,
                function_name
            ))
        })
        .collect();

    let mut modifiers = vec![];

    if adjoint {
        modifiers.push(GateModifier::Dagger);
    }

    if controlled {
        modifiers.push(GateModifier::Controlled);
    }

    let instruction = quil_rs::instruction::Instruction::Gate(quil_rs::instruction::Gate {
        name: name.to_owned(),
        parameters,
        qubits,
        modifiers,
    });

    pattern_context.quil_program.add_instruction(instruction);
}

lazy_static! {
    static ref QIS_INTRINSIC_REGEX: Regex = Regex::new(
        r"^__quantum__qis__(?P<operation>[^_]+)(?P<controlled>__ctl)?(?P<adjoint>__adj)?(__body)?$"
    )
    .unwrap();
}

#[allow(clippy::too_many_lines)]
pub(crate) fn quantum_instruction<'ctx>(
    context: &QCSCompilerContext<'ctx>,
    pattern_context: &mut ShotCountPatternMatchContext<'ctx>,
    instruction: InstructionValue<'ctx>,
) -> PatternResult<'ctx, ()> {
    match instruction.get_opcode() {
        inkwell::values::InstructionOpcode::Call => {
            let function_target_name = get_called_function_name(instruction);

            if let Some(function_name) = function_target_name {
                if let Some(captures) = QIS_INTRINSIC_REGEX.captures(&function_name) {
                    let operation = &captures["operation"];
                    let adjoint = captures.name("adjoint").is_some();
                    let controlled = captures.name("controlled").is_some();

                    let arguments = get_qis_function_arguments(context, instruction);

                    let matched = match operation {
                        "cnot" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "CNOT",
                                adjoint,
                                controlled,
                                0,
                                2,
                            );
                            true
                        }
                        "cz" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "CZ",
                                adjoint,
                                controlled,
                                0,
                                2,
                            );
                            true
                        }
                        "h" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "H",
                                adjoint,
                                controlled,
                                0,
                                1,
                            );
                            true
                        }
                        "reset" => {
                            // TODO: Alert the user that we're resetting for all qubits instead of just the targeted qubit here
                            pattern_context.use_active_reset = true;
                            true
                        }
                        "rz" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "RZ",
                                adjoint,
                                controlled,
                                1,
                                1,
                            );
                            true
                        }
                        "s" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "S",
                                adjoint,
                                controlled,
                                0,
                                1,
                            );
                            true
                        }
                        "x" => {
                            add_gate_instruction(
                                pattern_context,
                                &arguments,
                                &function_name,
                                "X",
                                adjoint,
                                controlled,
                                0,
                                1,
                            );
                            true
                        }
                        "mz" => {
                            let qubit =
                                *match_qis_argument!(Qubit, arguments, 0, function_name.as_str());
                            let result =
                                *match_qis_argument!(Result, arguments, 1, function_name.as_str());

                            // Result indices may be sparse rather than increasing monotonically from 0.
                            // If used naively (i.e. %Result 5 as `ro[5]`), this would result in sparse, suboptimal allocation
                            // in the readout data fetched following execution. So, instead, we assign Result indices to
                            // monotonically increasing `ro` region offsets so as to keep `ro` dense.
                            let next_ro_index = pattern_context.read_result_mapping.len() as u64;
                            let ro_buffer_index = pattern_context
                                .read_result_mapping
                                .entry(result)
                                .or_insert_with(|| next_ro_index);

                            pattern_context.quil_program.add_instruction(
                                quil_rs::instruction::Instruction::Measurement(
                                    quil_rs::instruction::Measurement {
                                        target: Some(MemoryReference {
                                            name: String::from("ro"),
                                            index: *ro_buffer_index,
                                        }),
                                        qubit: Qubit::Fixed(qubit),
                                    },
                                ),
                            );

                            true
                        }
                        _ => false,
                    };

                    if matched {
                        pattern_context.instructions_to_remove.push(instruction);
                        Some((instruction.get_next_instruction(), ()))
                    } else {
                        None
                    }
                } else if function_name == "__quantum__qis__read_result__body" {
                    let arguments = get_qis_function_arguments(context, instruction);
                    if let Some(OperationArgument::Result(result_index)) = arguments.get(0) {
                        let ro_index = pattern_context.read_result_mapping.get(result_index).unwrap_or_else(|| panic!("Result index {} was never the target of a measurement operation", result_index));
                        pattern_context
                            .readout_instruction_mapping
                            .push((*ro_index, instruction));
                    } else {
                        todo!("malformed read_result intrinsic")
                    }
                    pattern_context.instructions_to_remove.push(instruction);
                    Some((instruction.get_next_instruction(), ()))
                } else {
                    None
                }
            } else {
                None
            }
        }
        _ => None,
    }
}
