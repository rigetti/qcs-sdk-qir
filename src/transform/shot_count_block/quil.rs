// Copyright 2022 Rigetti Computing
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// This module is responsible for the transpilation of contiguous subsequences of LLVM instructions
// into quil, substituting those instructions with inline calls to a shared library responsible for
// executing those quil instructions.
use eyre::{eyre, Result};
use inkwell::{basic_block::BasicBlock, values::FunctionValue};
use quil_rs::instruction::Vector;

#[cfg(feature = "serde_support")]
use serde::{ser::SerializeStruct, Serialize, Serializer};

use crate::transform::PARAMETER_MEMORY_REGION_NAME;
use crate::{context::QCSCompilerContext, interop::entrypoint::get_entry_function, RecordedOutput};

use super::pattern::ShotCountPatternMatchContext;

/// Encapsulates the result of transpiling a QIR module to a Quil program
#[derive(Debug)]
pub struct ProgramOutput {
    /// The Quil program itself
    pub program: quil_rs::Program,
    /// The number of shots to run the program for, extracted from the primary execution loop
    pub shot_count: u64,
    /// Signifies output to be recorded at the end of program execution
    pub recorded_output: Vec<RecordedOutput>,
}

#[cfg(feature = "serde_support")]
impl Serialize for ProgramOutput {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut output = serializer.serialize_struct("ProgramOutput", 3)?;
        output.serialize_field("program", &self.program.to_string(true))?;
        output.serialize_field("shot_count", &self.shot_count)?;
        output.serialize_field("recorded_output", &self.recorded_output)?;
        output.end()
    }
}

/// Transform an entire QIR module to a single Quil program with shot count inferred
/// from a program loop counter.
#[allow(dead_code)]
pub(crate) fn transpile_module(context: &mut QCSCompilerContext) -> Result<ProgramOutput> {
    let entrypoint_function = get_entry_function(&context.module)
        .ok_or_else(|| eyre!("entrypoint not found in module"))?;
    transpile_function(context, entrypoint_function, &[])
}

/// Transpile a single QIR function body to a Quil program. This function may have any number
/// of basic blocks, but only the block named `body` will be parsed for quantum instructions
/// and transpiled to Quil; others will be ignored.
pub(crate) fn transpile_function<'ctx>(
    context: &mut QCSCompilerContext<'ctx>,
    function: FunctionValue<'ctx>,
    visited_functions: &[&str],
) -> eyre::Result<ProgramOutput> {
    let blocks = function.get_basic_blocks();
    let body_block = blocks
        .into_iter()
        .find(|el| el.get_name().to_string_lossy() == "body")
        .ok_or(eyre::eyre!("no basic block named 'body' found in function"))?;

    transpile_basic_block(context, body_block, visited_functions)
}

/// Transpile a single QIR basic block to a Quil program. This block must match the pattern
/// recognized by `ShotCountPatternMatchContext` in order to succeed.
pub(crate) fn transpile_basic_block<'ctx>(
    context: &mut QCSCompilerContext<'ctx>,
    basic_block: BasicBlock<'ctx>,
    visited_functions: &[&str],
) -> eyre::Result<ProgramOutput> {
    let pattern_context = ShotCountPatternMatchContext::from_basic_block(
        context,
        basic_block,
        visited_functions,
        fail_on_nested_function_call,
    )?;

    build_quil_program(context, &pattern_context)
}

/// This function exists to satisfy the signature needed to form a `ShotCountPatternMatchContext`.
///
/// However, because Quil does not itself have function calls, if there is a function call within
/// the basic block being transpiled, this returns an error.
pub(crate) fn fail_on_nested_function_call<'ctx>(
    _context: &mut QCSCompilerContext<'ctx>,
    function: FunctionValue<'ctx>,
    _visited_functions: &[&str],
) -> eyre::Result<()> {
    Err(eyre::eyre!(
        "cannot transpile nested function calls to Quil; found \"{}\"",
        function.get_name().to_string_lossy()
    ))
}

/// Build a Quil program from the information scraped into a shot-count pattern match.
/// If no pattern was detected, return an error.
pub(crate) fn build_quil_program<'ctx, 'p: 'ctx>(
    context: &mut QCSCompilerContext<'ctx>,
    pattern_context: &ShotCountPatternMatchContext<'p>,
) -> eyre::Result<ProgramOutput> {
    if let Some((program, shots)) = pattern_context.get_program_data() {
        let mut program = program.clone();

        program.add_instruction(quil_rs::instruction::Instruction::Declaration(
            quil_rs::instruction::Declaration {
                name: String::from("ro"),
                size: Vector {
                    data_type: quil_rs::instruction::ScalarType::Bit,
                    length: pattern_context.read_result_mapping.len() as u64,
                },
                sharing: None,
            },
        ));

        if !pattern_context.get_dynamic_parameters().is_empty() {
            program.add_instruction(quil_rs::instruction::Instruction::Declaration(
                quil_rs::instruction::Declaration {
                    name: String::from(PARAMETER_MEMORY_REGION_NAME),
                    size: Vector {
                        data_type: quil_rs::instruction::ScalarType::Real,
                        length: pattern_context.get_dynamic_parameters().len() as u64,
                    },
                    sharing: None,
                },
            ));
        }

        if pattern_context.use_active_reset {
            // Prepend a reset to the program via copy
            let instructions = program.to_instructions(true);
            let mut new_program = quil_rs::program::Program::new();
            new_program.add_instruction(quil_rs::instruction::Instruction::Reset(
                quil_rs::instruction::Reset { qubit: None },
            ));
            for instruction in instructions {
                new_program.add_instruction(instruction);
            }
            program = new_program;
        }

        if let Some(rewiring_pragma) = &context.options.rewiring_pragma {
            // Prepend a pragma to the program via copy
            let instructions = program.to_instructions(true);
            let mut new_program = quil_rs::program::Program::new();
            new_program.add_instruction(quil_rs::instruction::Instruction::Pragma(
                quil_rs::instruction::Pragma {
                    name: String::from("INITIAL_REWIRING"),
                    arguments: vec![format!("\"{}\"", rewiring_pragma.clone())],
                    data: None,
                },
            ));
            for instruction in instructions {
                new_program.add_instruction(instruction);
            }
            program = new_program;
        }

        Ok(ProgramOutput {
            program,
            shot_count: shots,
            recorded_output: pattern_context.recorded_output.clone(),
        })
    } else {
        Err(eyre::eyre!(
            "the shot count pattern was not detected in the program's basic block"
        ))
    }
}

#[cfg(test)]
mod test {
    use super::*;

    mod can_transpile_program_with {
        use super::*;
        use crate::context::context::{ContextOptions, QCSCompilerContext};
        use crate::context::target::ExecutionTarget;

        macro_rules! make_snapshot_test {
            ($name:ident) => {
                #[test]
                fn $name() {
                    let _ = env_logger::builder().is_test(true).try_init();

                    let base_context = inkwell::context::Context::create();
                    let data =
                        std::fs::read(format!("tests/fixtures/programs/{}.bc", stringify!($name)))
                            .unwrap();
                    let mut context = QCSCompilerContext::new_from_data(
                        &base_context,
                        &data,
                        ExecutionTarget::Qvm,
                        ContextOptions {
                            cache_executables: false,
                            rewiring_pragma: None,
                        },
                    )
                    .unwrap();
                    let result = transpile_module(&mut context).expect("transpilation failed");

                    insta::assert_snapshot!(result.program.to_string(true));
                }
            };
        }

        make_snapshot_test!(shot_count_loop);
        make_snapshot_test!(measure);
        make_snapshot_test!(measure_sparse);
        make_snapshot_test!(parametric);
        make_snapshot_test!(reduction);
        make_snapshot_test!(vqe_iteration);
        make_snapshot_test!(cartesian_rotations);
        make_snapshot_test!(pauli_xyz);
        make_snapshot_test!(s_and_adjoint_s);
        make_snapshot_test!(t_and_adjoint_t);
        make_snapshot_test!(toffoli);
        make_snapshot_test!(swap);
        make_snapshot_test!(entrypoint_attribute);
    }
}
