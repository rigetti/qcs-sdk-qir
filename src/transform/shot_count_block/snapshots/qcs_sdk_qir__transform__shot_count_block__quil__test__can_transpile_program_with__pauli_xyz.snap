---
source: src/transform/shot_count_block/quil.rs
expression: result.program.to_string(true)
---
DECLARE ro BIT[0]
X 0
Y 0
Z 0

